File: binV7/fourth_problem.rb
require 'json'

customers = [
  { "name": "john doe", "email": "john@example.com" },
  { "name": "John Doe", "email": "JOHN@example.com" },
  { "name": "jane smith", "email": "jane@example.com" }
]

grouped_customers = customers.group_by { |customer| [ customer[:name].downcase, customer[:email].downcase ] }

duplicates = grouped_customers.select { |k, v| v.size > 1 }.values.flatten

print JSON.pretty_generate({ data: duplicates })

File: binV15/fourth_problem.js
const customers = [
  { name: "john doe", email: "john@example.com" },
  { name: "John Doe", email: "JOHN@example.com" },
  { name: "jane smith", email: "jane@example.com" },
];

const groupedCustomers = Object.groupBy(customers, (customer) => [
  customer.name.toLowerCase(),
  customer.email.toLowerCase(),
]);

const duplicatedCustomers = Object.values(groupedCustomers).flatMap(
  (customer) => (customer.length > 1 ? customer : []),
);

console.log(JSON.stringify({ duplicatedCustomers }, null, 2));
